python/pythonUtilities.py

```diff
@@ -0,0 +1,177 @@
+        logMessage(f"  using command: {' '.join (commandArgs)}")
+
+    return processObject
+
+
+def getExistingClicastInstance(enviroPath):
+    """
+    This function will return the clicast instance for the given environment
+    or None if the environment does not have an instance, or if the process is not running
+    """
+
+    whatToReturn = None
+    # ensure a consistent key for the clicastInstances dictionary
+    enviroPath = cleanEnviroPath(enviroPath)
+    if enviroPath in clicastInstances and clicastInstances[enviroPath].poll() == None:
+        logMessage(
+            f"  using existing clicast instance [{clicastInstances[enviroPath].pid}] for: {enviroPath} "
+        )
+        whatToReturn = clicastInstances[enviroPath]
+
+    return whatToReturn
+
+
+def getClicastInstance(enviroPath):
+    """
+    This function will return the clicast instance for the given environment
+    If there is not an existing instance, a new clicast instance will be started.
+    """
+    clicastInstance = getExistingClicastInstance(enviroPath)
+    if clicastInstance == None:
+        clicastInstance = startNewClicastInstance(enviroPath)
+    return clicastInstance
+
+
+def closeEnvironmentConnection(enviroPath):
+    """
+    This function will terminate any clicast process that exists for enviroPath
+    It is used before things like delete and re-build environment, since we need
+    to delete the environment directory, and the running process will have it locked
+    If we are in server mode, we return True if we terminated a process, False otherwise
+    """
+
+    returnValue = False
+    if USE_SERVER:
+        processObject = getExistingClicastInstance(enviroPath)
+        if processObject != None:
+            logMessage(
+                f"  terminating clicast instance [{processObject.pid}] for environment: {enviroPath}"
+            )
+            # This tells clicast to shutdown gracefully
+            # In the case where the server has been stopped with ctrl-c
+            # we get here, but the clicast process might have already died
+            # from the propagated SIGINT, so we need to catch the exception
+            try:
+                processObject.stdin.write("clicast-server-shutdown\n")
+                processObject.stdin.flush()
+                processObject.wait()
+            except:
+                pass
+            # This simply removes the processObject from the dictionary
+            removeClicastInstance(enviroPath)
+            returnValue = True
+        else:
+            logMessage(f"  no clicast instance exists for environment: {enviroPath}")
+
+    return returnValue
+
+
+def cleanEnviroPath(enviroPath):
+    """
+    This function is used to clean up the environment path to make it usable as
+    a consistent dictionary key.  We force the drive lever to lower case
+    and we replace backslashes with forward slashes.
+    """
+    returnPath = enviroPath.replace("\\", "/")
+    if returnPath[2] == ":":
```

Comment:
I guess we'd want to check `returnPath[1]` for stuff like `C:/path/to/some/env`

Reply:
@Zbigor Great catch on this one, thanks!
Fixed with: 817e1dc